{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Before beginning...\n",
    "I execute:\n",
    "* In a terminal: docker start course-mysql\n",
    "* the connection lines in this jupyter notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 1 \n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Firstly, I execute:\n",
    "* import re to bringing the python regular expression object into this jupyter notebook.\n",
    "* import pymysql.cursors to loading the library package which allows interaction Python-MySQL.\n",
    "\n",
    "Before, I open two files (Germplasm.tsv and LocusGene.tsv) for reading. Also, I open two files with Mozilla Firefox for watching its structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pymysql.cursors\n",
    "\n",
    "Germplasm = open('Germplasm.tsv', 'r')\n",
    "LocusGene = open('LocusGene.tsv', 'r')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Explanation** \n",
    "\n",
    "Firstly, I have used seek(0) command to reset back to the beginning, and that reading all file.\n",
    "\n",
    "Secondly, I have created two lists to collect only the information from AGI Locus Code since according to hypothetical scenario, the two files contain the same information in that column.\n",
    "\n",
    "With conditionals, I have selected target values of lists (AGI Locus Code column) to compare if both are the same.\n",
    "\n",
    "For comparison, two files have to interact. With \"for\" command I have succeeded. When I opened the files, I saw that they were tables and the Locus AGI Code were arranged in the same order. \n",
    "\n",
    "I have created a loop to read each row from the Germplasm table. With readlines(), I can read row by row one-at-a-time and do something useful with each read.Before, I have looked for the AGI which has the following structure: ATxGxxxxxx. I have used re.search command. Then, I have written a condition, If there is a match, I save those values. And this is repeated for the LocusGene file. Previously, I have used seek(0) command to reset back to the beginning. \n",
    "\n",
    "Finally, with comparison operators (==), I did the comparison between the two files that contained all the Locus AGI Code, and also, if files didn't match each other, I set its to tell me that there wasn't match (No match!!). On the other hand, if files matched each order, I set its to tell me Locus AGI Code + common value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No match!!\n",
      "AT1G01040\t common values\n",
      "AT1G01060\t common values\n",
      "AT1G01140\t common values\n",
      "AT1G01220\t common values\n",
      "AT2G03720\t common values\n",
      "AT2G03800\t common values\n",
      "AT2G04240\t common values\n",
      "AT2G05210\t common values\n",
      "AT3G02130\t common values\n",
      "AT3G02140\t common values\n",
      "AT3G02230\t common values\n",
      "AT3G02260\t common values\n",
      "AT3G02310\t common values\n",
      "AT3G02680\t common values\n",
      "AT3G02850\t common values\n",
      "AT3G02870\t common values\n",
      "AT3G03260\t common values\n",
      "AT4G14790\t common values\n",
      "AT4G15210\t common values\n",
      "AT4G15560\t common values\n",
      "AT4G15570\t common values\n",
      "AT4G15802\t common values\n",
      "AT4G15880\t common values\n",
      "AT4G16420\t common values\n",
      "AT4G16480\t common values\n",
      "AT5G10480\t common values\n",
      "AT5G10510\t common values\n",
      "AT5G11110\t common values\n",
      "AT5G11260\t common values\n",
      "AT5G11510\t common values\n",
      "AT5G12200\t common values\n",
      "AT5G13290\t common values\n"
     ]
    }
   ],
   "source": [
    "Germplasm.seek(0)\n",
    "GermplasmAGICode = []\n",
    "LocusGeneAGICode = []\n",
    "\n",
    "for row in Germplasm.readlines():\n",
    "    match = re.search (r'AT\\dG\\d+\\s', row)\n",
    "    if match:\n",
    "        GermplasmRowCode = match.group()\n",
    "        GermplasmAGICode.append(GermplasmRowCode)\n",
    "        LocusGene.seek(0)\n",
    "        for row2 in LocusGene.readlines():\n",
    "            match2 = re.search(r'AT\\dG\\d+\\s', row2)\n",
    "            if match2:\n",
    "                LocusGeneRowCode=match2.group()\n",
    "                if (GermplasmRowCode)==(LocusGeneRowCode):\n",
    "                    LocusGeneAGICode.append(LocusGeneRowCode)\n",
    "                    print(LocusGeneRowCode, 'common values')\n",
    "    else:\n",
    "        print('No match!!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 2: Design and create the database.\n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* You may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Explanation**\n",
    "\n",
    "Firstly, I have created database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql drop database DatabaseExam2;\n",
    "#%sql show databases;\n",
    "#I did it for trying to get back the values in tables, but it wasn't working for LocusGene table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "17 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>DatabaseExam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>DatabaseExam2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LocusGeneexam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Organisms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>OrganismsEasier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProyectEasier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProyectEasier2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Researcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ResearcherEasier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasmexam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>testing123</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('DatabaseExam',),\n",
       " ('DatabaseExam2',),\n",
       " ('LocusGeneexam',),\n",
       " ('Organisms',),\n",
       " ('OrganismsEasier',),\n",
       " ('ProyectEasier',),\n",
       " ('ProyectEasier2',),\n",
       " ('Researcher',),\n",
       " ('ResearcherEasier',),\n",
       " ('Samples',),\n",
       " ('germplasm',),\n",
       " ('germplasmexam',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',),\n",
       " ('testing123',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database DatabaseExam2;\n",
    "%sql show databases;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use DatabaseExam2;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Secondly, I have created tables. The syntax of create table:\n",
    "* Germplasm Table:\n",
    "\n",
    "-LocusGermplasm VARCHAR(15) NOT NULL PRIMARY KEY: VARCHAR because it can be any character, and NOT NULL because it always has to have some value. The limit is 15 to be sure although we know that it is 9 characters.\n",
    "\n",
    "The following I haven't written NOT NULL because they may be empty since it is an unknown file and I don't know if there are values in all the fields. I have written that name so as not to be confused with the Locus column of the LocusGene.tsv file.\n",
    "\n",
    "-germplasm VARCHAR(25): VARCHAR because it can be any character.\n",
    "\n",
    "-phenotype VARCHAR(2000): VARCHAR because it can be any character. The limit is 20 to be sure because this datatype is description.\n",
    "\n",
    "-pubmed INTEGER. INTEGER because it can be any number.\n",
    "\n",
    "* LocusGene Table:\n",
    "-Locus_id VARCHAR(15) NOT NULL PRIMARY KEY: VARCHAR because it can be any character, and NOT NULL because it always has to have some value. The limit is 15 to be sure although we know that it is 9 characters.\n",
    "\n",
    "The following I haven't written NOT NULL because they may be empty since it is an unknown file and I don't know if there are values in all the fields. I have written that name so as not to be confused with the Locus column of the LocusGene.tsv file.\n",
    "\n",
    "-gene VARCHAR(25): VARCHAR because it can be any character.\n",
    "\n",
    "-protein_length INTEGER. INTEGER because it can be any number.\n",
    "\n",
    "\n",
    "I have written PRIMARY KEY in LocusGermplasm and Locus_id because both columns are identical, as I have tested in Problem 1. Therefore, both tables can be joined by these columns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LocusGermplasm</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(25)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(2000)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('LocusGermplasm', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(25)', 'YES', '', None, ''),\n",
       " ('phenotype', 'varchar(2000)', 'YES', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'YES', '', None, '')]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE Germplasm(LocusGermplasm VARCHAR(15) NOT NULL PRIMARY KEY, germplasm VARCHAR(25), phenotype VARCHAR(2000), pubmed INTEGER);\n",
    "%sql DESCRIBE Germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus_id</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(25)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>protein_length</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>YES</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('Locus_id', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('gene', 'varchar(25)', 'YES', '', None, ''),\n",
       " ('protein_length', 'int(11)', 'YES', '', None, '')]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE LocusGene(Locus_id VARCHAR(15) NOT NULL PRIMARY KEY, gene VARCHAR(25), protein_length INTEGER);\n",
    "%sql DESCRIBE LocusGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>LocusGermplasm</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>Locus_id</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM Germplasm INNER JOIN LocusGene ON LocusGermplasm = Locus_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 3: Fill the database\n",
    "* Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Explanation**\n",
    "Firstly, I have connected to the database. I have used the command for this network and I have specified this database (db='DatabaseExam2').\n",
    "\n",
    "Secondly, I try to loading database. I have open the file. I know that the files have the columns separated by a tab (delimiter = '\\t') and the fields (columns). So, I have created variable (GermplasmReader or LocusGeneReader). Then, I have loaded data with INSERT INTO and VALUES, defining its format. In this part, it is important to remember that the pubmed and protein legth columns are defined only for digits. For this reason they are not added ''.\n",
    "\n",
    "I have used SELECT * FROM Germplasm for checking that values appear in tables. I have a problem with LocusGene and I don't know why I can't see values...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('Locus', 'Locus'), ('germplasm', 'germplasm'), ('phenotype', 'phenotype'), ('pubmed', 'pubmed')])\n",
      "OrderedDict([('Locus', 'AT1G01040'), ('germplasm', 'CS3828'), ('phenotype', 'Increased abundance of miRNA precursors.'), ('pubmed', '17369351')])\n",
      "OrderedDict([('Locus', 'AT1G01060'), ('germplasm', 'lhy-101'), ('phenotype', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.'), ('pubmed', '16891401')])\n",
      "OrderedDict([('Locus', 'AT1G01140'), ('germplasm', 'SALK_058629'), ('phenotype', 'hypersensitive to low potassium media'), ('pubmed', '17486125')])\n",
      "OrderedDict([('Locus', 'AT1G01220'), ('germplasm', 'SALK_012400C'), ('phenotype', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.'), ('pubmed', '18199744')])\n",
      "OrderedDict([('Locus', 'AT2G03720'), ('germplasm', 'SALK_042433'), ('phenotype', 'Multiple straight hairs'), ('pubmed', '16367956')])\n",
      "OrderedDict([('Locus', 'AT2G03800'), ('germplasm', 'gek1-1'), ('phenotype', 'Ethanol hypersensitivity.'), ('pubmed', '15215505')])\n",
      "OrderedDict([('Locus', 'AT2G04240'), ('germplasm', 'xerico'), ('phenotype', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.'), ('pubmed', '17933900')])\n",
      "OrderedDict([('Locus', 'AT2G05210'), ('germplasm', 'pot1-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '17627276')])\n",
      "OrderedDict([('Locus', 'AT3G02130'), ('germplasm', 'rpk2-2'), ('phenotype', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.'), ('pubmed', '17419837')])\n",
      "OrderedDict([('Locus', 'AT3G02140'), ('germplasm', 'afp4-1'), ('phenotype', 'Decreased germination on high concentrations of glucose and sorbitol.'), ('pubmed', '18484180')])\n",
      "OrderedDict([('Locus', 'AT3G02230'), ('germplasm', 'rgp1-1'), ('phenotype', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.'), ('pubmed', '21478444')])\n",
      "OrderedDict([('Locus', 'AT3G02260'), ('germplasm', 'tir3-1 RGLG1:rglg1 rglg2'), ('phenotype', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.'), ('pubmed', '17586653')])\n",
      "OrderedDict([('Locus', 'AT3G02310'), ('germplasm', 'sep2-1'), ('phenotype', 'Non-described subtle phenotype.'), ('pubmed', '10821278')])\n",
      "OrderedDict([('Locus', 'AT3G02680'), ('germplasm', 'atnbs1-1'), ('phenotype', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.'), ('pubmed', '17672843')])\n",
      "OrderedDict([('Locus', 'AT3G02850'), ('germplasm', 'CS3816'), ('phenotype', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.'), ('pubmed', '17568770')])\n",
      "OrderedDict([('Locus', 'AT3G02870'), ('germplasm', 'vtc4-1'), ('phenotype', 'ascorbate deficient'), ('pubmed', '16595667')])\n",
      "OrderedDict([('Locus', 'AT3G03260'), ('germplasm', 'hdg8-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '16778018')])\n",
      "OrderedDict([('Locus', 'AT4G14790'), ('germplasm', 'pdd17'), ('phenotype', 'Defective pollen development.'), ('pubmed', '19237690')])\n",
      "OrderedDict([('Locus', 'AT4G15210'), ('germplasm', 'bmy1-2'), ('phenotype', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.'), ('pubmed', '16297066')])\n",
      "OrderedDict([('Locus', 'AT4G15560'), ('germplasm', 'cla1-1'), ('phenotype', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.'), ('pubmed', '10982425')])\n",
      "OrderedDict([('Locus', 'AT4G15570'), ('germplasm', 'maa3'), ('phenotype', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.'), ('pubmed', '18772186')])\n",
      "OrderedDict([('Locus', 'AT4G15802'), ('germplasm', 'Athspb-2'), ('phenotype', 'Early flowering, reduced fertility, aborted seeds.'), ('pubmed', '20388662')])\n",
      "OrderedDict([('Locus', 'AT4G15880'), ('germplasm', 'esd4-2'), ('phenotype', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.'), ('pubmed', '17513499')])\n",
      "OrderedDict([('Locus', 'AT4G16420'), ('germplasm', 'prz1-1'), ('phenotype', 'Altered response to auxin and cytokinin'), ('pubmed', '12747832')])\n",
      "OrderedDict([('Locus', 'AT4G16480'), ('germplasm', 'atint4-2'), ('phenotype', 'No visible phenotype.'), ('pubmed', '16603666')])\n",
      "OrderedDict([('Locus', 'AT5G10480'), ('germplasm', 'pas2-3'), ('phenotype', 'Segregates 25% embryo lethal.'), ('pubmed', '18799749')])\n",
      "OrderedDict([('Locus', 'AT5G10510'), ('germplasm', 'plt3-1'), ('phenotype', 'Short roots and shortened root meristem.'), ('pubmed', '17960244')])\n",
      "OrderedDict([('Locus', 'AT5G11110'), ('germplasm', 'kns2'), ('phenotype', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.'), ('pubmed', '18779216')])\n",
      "OrderedDict([('Locus', 'AT5G11260'), ('germplasm', 'hy5-101'), ('phenotype', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.'), ('pubmed', '16891401')])\n",
      "OrderedDict([('Locus', 'AT5G11510'), ('germplasm', 'myb3r4-1'), ('phenotype', 'No visible phenotype.'), ('pubmed', '17287251')])\n",
      "OrderedDict([('Locus', 'AT5G12200'), ('germplasm', 'pyd2-2'), ('phenotype', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.'), ('pubmed', '19413687')])\n",
      "OrderedDict([('Locus', 'AT5G13290'), ('germplasm', 'crn-1'), ('phenotype', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.'), ('pubmed', '12345678')])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pymysql.cursors\n",
    "\n",
    "import csv\n",
    "import io\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                            user='root',\n",
    "                            password='root',\n",
    "                            db='DatabaseExam2',\n",
    "                            charset='utf8mb4',\n",
    "                            cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open ('Germplasm.tsv') as csvfile:\n",
    "            GermplasmReader = csv.DictReader(csvfile, delimiter = '\\t', fieldnames = (\"Locus\", \"germplasm\", \"phenotype\", \"pubmed\"))\n",
    "            for row in GermplasmReader:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO Germplasm(LocusGermplasm, germplasm, phenotype, pubmed) VALUES\n",
    "                ('{}', '{}', '{}', {})\"\"\".format (row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "    connection.commit()\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM Germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "33 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>LocusGermplasm</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>germplasm</td>\n",
       "        <td>phenotype</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678),\n",
       " ('Locus', 'germplasm', 'phenotype', None)]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use DatabaseExam2;\n",
    "%sql SELECT * FROM Germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('Locus', 'Locus'), ('Gene', 'Gene'), ('ProteinLength', 'ProteinLength')])\n",
      "OrderedDict([('Locus', 'AT1G01040'), ('Gene', 'DCL1'), ('ProteinLength', '332')])\n",
      "OrderedDict([('Locus', 'AT1G01060'), ('Gene', 'LHY'), ('ProteinLength', '290')])\n",
      "OrderedDict([('Locus', 'AT1G01140'), ('Gene', 'CIPK9'), ('ProteinLength', '223')])\n",
      "OrderedDict([('Locus', 'AT1G01220'), ('Gene', 'FKGP'), ('ProteinLength', '190')])\n",
      "OrderedDict([('Locus', 'AT2G03720'), ('Gene', 'MRH6'), ('ProteinLength', '189')])\n",
      "OrderedDict([('Locus', 'AT2G03800'), ('Gene', 'GEK1'), ('ProteinLength', '196')])\n",
      "OrderedDict([('Locus', 'AT2G04240'), ('Gene', 'XERICO'), ('ProteinLength', '256')])\n",
      "OrderedDict([('Locus', 'AT2G05210'), ('Gene', 'POT1A'), ('ProteinLength', '221')])\n",
      "OrderedDict([('Locus', 'AT3G02130'), ('Gene', 'RPK2'), ('ProteinLength', '284')])\n",
      "OrderedDict([('Locus', 'AT3G02140'), ('Gene', 'TMAC2'), ('ProteinLength', '300')])\n",
      "OrderedDict([('Locus', 'AT3G02230'), ('Gene', 'RGP1'), ('ProteinLength', '301')])\n",
      "OrderedDict([('Locus', 'AT3G02260'), ('Gene', 'BIG'), ('ProteinLength', '279')])\n",
      "OrderedDict([('Locus', 'AT3G02310'), ('Gene', 'SEP2'), ('ProteinLength', '175')])\n",
      "OrderedDict([('Locus', 'AT3G02680'), ('Gene', 'NBS1'), ('ProteinLength', '190')])\n",
      "OrderedDict([('Locus', 'AT3G02850'), ('Gene', 'SKOR'), ('ProteinLength', '234')])\n",
      "OrderedDict([('Locus', 'AT3G02870'), ('Gene', 'VTC4'), ('ProteinLength', '311')])\n",
      "OrderedDict([('Locus', 'AT3G03260'), ('Gene', 'HDG8'), ('ProteinLength', '194')])\n",
      "OrderedDict([('Locus', 'AT4G14790'), ('Gene', 'SUV3'), ('ProteinLength', '312')])\n",
      "OrderedDict([('Locus', 'AT4G15210'), ('Gene', 'BAM5'), ('ProteinLength', '313')])\n",
      "OrderedDict([('Locus', 'AT4G15560'), ('Gene', 'DXS'), ('ProteinLength', '219')])\n",
      "OrderedDict([('Locus', 'AT4G15570'), ('Gene', 'MAA3'), ('ProteinLength', '294')])\n",
      "OrderedDict([('Locus', 'AT4G15802'), ('Gene', 'HSBP'), ('ProteinLength', '254')])\n",
      "OrderedDict([('Locus', 'AT4G15880'), ('Gene', 'ESD4'), ('ProteinLength', '265')])\n",
      "OrderedDict([('Locus', 'AT4G16420'), ('Gene', 'ADA2B'), ('ProteinLength', '279')])\n",
      "OrderedDict([('Locus', 'AT4G16480'), ('Gene', 'INT4'), ('ProteinLength', '284')])\n",
      "OrderedDict([('Locus', 'AT5G10480'), ('Gene', 'PAS2'), ('ProteinLength', '301')])\n",
      "OrderedDict([('Locus', 'AT5G10510'), ('Gene', 'AIL6'), ('ProteinLength', '310')])\n",
      "OrderedDict([('Locus', 'AT5G11110'), ('Gene', 'SPS2'), ('ProteinLength', '232')])\n",
      "OrderedDict([('Locus', 'AT5G11260'), ('Gene', 'HY5'), ('ProteinLength', '221')])\n",
      "OrderedDict([('Locus', 'AT5G11510'), ('Gene', 'MYB3R-4'), ('ProteinLength', '336')])\n",
      "OrderedDict([('Locus', 'AT5G12200'), ('Gene', 'PYD2'), ('ProteinLength', '310')])\n",
      "OrderedDict([('Locus', 'AT5G13290'), ('Gene', 'CRN'), ('ProteinLength', '189')])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pymysql.cursors\n",
    "\n",
    "import csv\n",
    "import io\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                            user='root',\n",
    "                            password='root',\n",
    "                            db='DatabaseExam2',\n",
    "                            charset='utf8mb4',\n",
    "                            cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open ('LocusGene.tsv') as csvfile:\n",
    "            LocusGeneReader = csv.DictReader(csvfile, delimiter = '\\t', fieldnames = (\"Locus\", \"Gene\", \"ProteinLength\"))\n",
    "            for row in LocusGeneReader:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO LocusGene(Locus_id, gene, protein_length) VALUES\n",
    "                ('{}', '{}', {})\"\"\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "                #cursor.execute(sql)\n",
    "    connection.commit()\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus_id</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use DatabaseExam2;\n",
    "%sql SELECT * FROM LocusGene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 4: Create reports, written to a file\n",
    "* Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "* Create a joined report that only includes the Genes SKOR and MAA3\n",
    "* Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "* Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a report that shows the full, joined, content of the two database tables (including a header line):\n",
    "\n",
    "**Explanation**\n",
    "* For querying the common data, I have used SELECT * FROM and I connected both tables for Locus row. In problem 2, I have written PRIMARY KEY in LocusGermplasm and Locus_id because both columns are identical, as I have tested in Problem 1. Therefore, both tables can be joined by these columns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>LocusGermplasm</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>Locus_id</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM Germplasm, LocusGene WHERE \\\n",
    "     LocusGermplasm = Locus_id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a joined report that only includes the Genes SKOR and MAA3.\n",
    "\n",
    "**Explanation**\n",
    "\n",
    "I have used INNER JOIN to connect both tables with locus row. Then, I have written WHERE to  conditionally select data because I just want the column gene, and also I have used LIKE for searching specific genes: SKOR and MAA3.\n",
    "\n",
    "I have write str to returns the string version of the given object and I have printed. I don't know why the values of LocusGene table don't appear... So I can't see the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      "+----------------+-----------+-----------+--------+----------+------+----------------+\n",
      "| LocusGermplasm | germplasm | phenotype | pubmed | Locus_id | gene | protein_length |\n",
      "+----------------+-----------+-----------+--------+----------+------+----------------+\n",
      "+----------------+-----------+-----------+--------+----------+------+----------------+\n"
     ]
    }
   ],
   "source": [
    "import re \n",
    "\n",
    "\n",
    "Report = %sql SELECT * FROM Germplasm INNER JOIN LocusGene ON LocusGermplasm = Locus_id WHERE gene LIKE 'SOR' OR gene LIKE 'MAA3'\n",
    "Report = str(Report)\n",
    "print(Report)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "+-------------------+\n",
      "| Number of entries |\n",
      "+-------------------+\n",
      "|         0         |\n",
      "+-------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "+-------------------+\n",
      "| Number of entries |\n",
      "+-------------------+\n",
      "|         0         |\n",
      "+-------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "+-------------------+\n",
      "| Number of entries |\n",
      "+-------------------+\n",
      "|         0         |\n",
      "+-------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "+-------------------+\n",
      "| Number of entries |\n",
      "+-------------------+\n",
      "|         0         |\n",
      "+-------------------+\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      "+-------------------+\n",
      "| Number of entries |\n",
      "+-------------------+\n",
      "|         0         |\n",
      "+-------------------+\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "Report1 = %sql SELECT COUNT(*) AS \"Number of entries for Chromosome 1\" FROM Germplasm RIGHT JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE 'AT1\\dG\\d+\\s';\n",
    "Report1 = str(Report1)\n",
    "print(Report1)\n",
    "\n",
    "Report2 = %sql SELECT COUNT(*) AS \"Number of entries for Chromosome 2\" FROM Germplasm RIGHT JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE 'AT2\\dG\\d+\\s';\n",
    "Report2 = str(Report2)\n",
    "print(Report2)\n",
    "\n",
    "Report3 = %sql SELECT COUNT(*) AS \"Number of entries for Chromosome 3\" FROM Germplasm RIGHT JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE 'AT3\\dG\\d+\\s';\n",
    "Report3 = str(Report3)\n",
    "print(Report3)\n",
    "\n",
    "Report4 = %sql SELECT COUNT(*) AS \"Number of entries for Chromosome 4\" FROM Germplasm RIGHT JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE 'AT4\\dG\\d+\\s';\n",
    "Report4 = str(Report4)\n",
    "print(Report4)\n",
    "\n",
    "Report5 = %sql SELECT COUNT(*) AS \"Number of entries for Chromosome 5\" FROM Germplasm RIGHT JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE 'AT5\\dG\\d+\\s';\n",
    "Report5 = str(Report5)\n",
    "print(Report5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "Report6 = %sql SELECT AVG(LocusGene.protein_length) FROM Germplasm INNER JOIN LocusGene ON LocusGermplasm = Locus_id WHERE LocusGermplasm LIKE '\\sddd\\';\n",
    "Report6 = str(Report6)\n",
    "print(Report6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Explanation** I don't know why because I haven't values in LocusGene table. I haven't been able to separate the data you were asking for. In the following lines, I explain how I would have made the report:\n",
    "\n",
    "Firstly, I connect with database:\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                            user='root',\n",
    "                            password='root',\n",
    "                            db='DatabaseExam2',\n",
    "                            charset='utf8mb4',\n",
    "                            cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "After, I would have written this:\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "    sql=\"\"\"SELECT ... (depending on each question)\"\"\"\n",
    "    cursor.execute(sql)\n",
    "    results1 = cursor.fetchall()  [for bringing back all the rows of a query result]\n",
    "    Report = open ('Report.txt', 'w')\n",
    "    Report.write (depending on each question)\n",
    "    \n",
    "  \n",
    "  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
